feat(algebra/algebra/non_unital): define non-unital, non-associative algebras

ocfnash opened the pull request:
With sufficient refactoring effort, the `non_unital_non_assoc_algebra` proposed here could replace our extant `algebra` definition entirely. This would have the advantage that whether we wish to specialize to unital / associative algebras would then be controlled entirely by whether we pass a `non_assoc_semiring` / `non_unital_semiring`  / ... as the unbundled typeclass.

The extant `algebra` definition uses both the existence of units and associativity. However the path to remove its dependence on associativity looks fairly easy since this arises through the redundant requirement of associativity in `ring_hom` (via `monoid_hom`) in the usual way: an overly proscriptive typeclass. I actually explored relaxing this (i.e., changing `[monoid M]` to `[has_mul M] [has_one M]` in the definition of `monoid_hom`) but I quickly ran into difficulties associated with `to_additive` magic.

I also experimented with using the `mul_one_class` and `add_zero_class` in the definitions of `monoid` and `add_monoid` and ran into trouble. However I think this could be resolved fairly easily. It's just a slow debug loop because of the long build time. Since the gain from such a change is negligible I gave up.

Co-authored-by: Eric Wieser wieser.eric@gmail.com

---

For what it's worth, I'm defining these with the following plan in mind:

1. Define non-unital, non-assoc algebras
2. Construct the magma algebra (I claim this needs non-unital, non-assoc algebras to define universal property conveniently)
3. Construct the free non-associative algebra as the magma algebra of the free magma
4. Construct the free Lie algebra as a quotient of the free non-associative algebra
5. Construct the semisimple Lie algebras as quotients of the free Lie algebra using the Serre-Chevalley construction (with relations defined by their Cartan matrices)
6. Define the exceptional Lie algebras via the Cartan matrices

There is now a branch `replace_algebra_def` that builds off this PR and attempts to replace the definition of `algebra` as described. It is still incomplete.

- [x] depends on: #6786

ocfnash committed:
feat(algebra/non_associative/basic): define non-associative algebras

ocfnash commented on src/algebra/non_associative/basic.lean:
```diff
@@ -0,0 +1,155 @@
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
```
Note that I don't assume `R` is commutative. Perhaps I should?

For unital algebras we always assume commutativity since any action factors through the commutative quotient but I don't see that this is the case here.

ocfnash commented on src/algebra/non_associative/basic.lean:
```diff
@@ -0,0 +1,155 @@
+def inverse (f : na_algebra_hom R A B) (g : B → A)
+  (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f) : na_algebra_hom R B A :=
+{ map_mul' := λ x y,
+  calc g (x * y) = g ((f (g x)) * (f (g y))) : by { conv_lhs { rw [←h₂ x, ←h₂ y], }, }
```
I guess this might be better factored out into a result about `mul_hom`s. Since I'm not yet sure if this PR will be accepted, I thought I'd leave things like this for now to keep it simple.

eric-wieser commented on src/algebra/non_associative/basic.lean:
```diff
@@ -0,0 +1,155 @@
+set_option old_structure_cmd true
+
+/-- A morphism of non-associative algebras. -/
+structure na_algebra_hom extends A →ₗ[R] B, mul_hom A B
```
Strictly this is just about non-unital algebras, right? You'd use the same type whether or not the algebra is associative.

ocfnash replied:
Yes, quite right. At the very least, I should reword the comment. I'll wait and see what others think before I start updating names though since I'm fairly sure that `na_algebra` won't survive!

Julian-Kuelshammer replied:
Would `nassoc_algebra` be too long?

eric-wieser replied:
I'd probably lean towards `pre_algebra`, since the only reason we don't want to use this as `algebra` is definitional - the axiomatic content is equivalent in all the places that `algebra` is used today.

Julian-Kuelshammer replied:
I don't like the idea of introducing new terminology that everyone has to look up. According to Wikipedia, non-associative or distributive are used: https://en.m.wikipedia.org/wiki/Non-associative_algebra

eric-wieser replied:
@Julian-Kuelshammer, my point is that for "non-associativity", we may as well just redefine `algebra` to require `na_semiring` instead of `semiring`. Existing lemmas that assume `semiring` will keep working.

It's only non-unitarity that results in needing a separate class.

Julian-Kuelshammer replied:
I think it is quite common in mathematics to study non-associative algebras over commutative associative (semi)rings. That's not too far from my own area of research. I have not seen algebras over non-associative rings (but that might be a fault in my perspective).

ocfnash replied:
@Julian-Kuelshammer I don't think @eric-wieser is suggesting having non-associative coefficients (I too find this very exotic, though perhaps hard core groups theorists study Lie ring modules?).

Ignoring the detail that I have proposed allowing non-commutative coefficients for non-assoc algebras, I think Eric is pointing out that we don't really need to have both `algebra` and `na_algebra` classes since they could have the same definition and differ only in whether they consume a`[na_semiring A]` or a `[semiring A]` class.

I wholeheartedly agree with this point but I worry that the required refactor (changing the extant definition of `algebra`) would stall my efforts to get non-associative algebras off the ground. I thus advocate the approach proposed here: get started with `na_algebra` and refactor later.

Oh and on naming, I like `nassoc_algebra` best.

eric-wieser replied:
@ocfnash characterizes my stance well, although it's a bit more nuanced right now. Using "nn" for "not necesarily"

* If we want unitary-but-nn-associative `algebra`s, then by far the easiest way to proceed is to just weaken the typeclass assumptions on the current `algebra`
* If we want nn-unitary-or-associative `algebra`s, then there's a can of worms that's probably worth avoiding for now, and making the new typeclass is reasonable.

Given the fact that the unitarity is what motivates the new typeclass, I'd argue it should be called `nunit_algebra` or `nonunital_algebra`; because otherwise, after making the refactor for unitary-but-nn-associative algebras, we'd have `nassoc_algebra` which is nn-unital and nn-associative and `algebra` which is nn-associative, which is just confusing!

ocfnash replied:
In the end I have opted for the following names:

- `non_unital_non_assoc_semiring`
- `non_unital_semiring`
- `non_assoc_semiring`
- `non_unital_non_assoc_algebra`
- `non_unital_non_assoc_algebra_hom`

eric-wieser commented on src/algebra/non_associative/basic.lean:
```diff
@@ -0,0 +1,155 @@
+/-- A not-necessarily-unital, not-necessarily-associative algebra, or "non-associative algebra" for
+short. -/
+@[protect_proj]
+class na_algebra [na_semiring A] extends semimodule R A :=
```
If `A` is a `comm_semiring`, are these axioms sufficient to reconstruct an `algebra`? That is to say, what assumptions do you need to define `na_algebra.to_algebra`?

ocfnash replied:
To define `na_algebra.to_algebra` we need:

- Commutative coefficients `R`
- A unit in `A`
- Associative multiplication in `A`

eric-wieser replied:
`[comm_semiring A]` provides the last two, and I think the axioms are at least enough to show that `(s * t) • a = (t * s) • a`, but I guess that's not quite as strong as showing that `s * t = t * s`

ocfnash replied:
I think what you are thinking of might be:
```lean
def na_algebra.to_algebra
  (R A : Type*) [comm_semiring R] [semiring A] [na_algebra R A] : algebra R A :=
sorry
```

eric-wieser replied:
That's certainly one of the things I was thinking, yes. At that point; there's only one difference between `[comm_semiring R] [semiring A] [na_algebra R A]` and `[comm_semiring R] [semiring A] [algebra R A]`; the ability to control the definitional reduction of `algebra_map`.

ocfnash replied:
I've added a commit along these lines.

ocfnash committed:
Add `na_algebra.to_algebra`

eric-wieser commented on src/algebra/non_associative/basic.lean:
```diff
@@ -0,0 +1,160 @@
+
+/-- A not-necessarily-unital, not-necessarily-associative semiring, or "non-associative ring" for
+short. -/
+class na_semiring extends add_comm_monoid A, distrib A, mul_zero_class A
```
We could at this point just open the floodgates and add all the missing ones:

```suggestion
class mul_one_class extends has_one, has_mul :=
(one_mul : ∀ (a : M), 1 * a = a)
(mul_one : ∀ (a : M), a * 1 = a)
class semigroup_with_zero extends semigroup A, mul_zero_class A
class nau_semiring extends add_comm_monoid A, distrib A, mul_zero_class A
class nu_semiring extends nau_semiring A, semigroup_with_zero A
class na_semiring extends nau_semiring A, mul_one_class A
```

Then the existing `algebra` definition could happily require `na_semiring` with minimal breakage. Supporting `nau_semiring` requires the new class you add in this PR.

ocfnash replied:
This may well be the right approach. For one thing, we will eventually want to talk about unital non-assoc algebras.

I'm biased to push ahead the the MVP and maybe return to this but perhaps now is the time.

ocfnash committed:
Remove simp attributes to appease linter

ocfnash committed:
Factor out `mul_hom.inverse`

ocfnash committed:
Experiment with Eric's "open the floodgates" suggestion

Despite initial reluctance, I am beginning to be persuaded. The
names are not serious suggestions, especially as if this works, what
we are really defining is non-unital algebras since we can make the
existing `algebra` definition allow for non-associativity for free.

For now I just want to see whether this compiles.

ocfnash committed:
Fix build failure 1 / ?

ocfnash committed:
Change `semiring` ancestors.

_Probably_ a more useful set of ancestors.

ocfnash committed:
Revert `semiring` ancestors change.

For reasons I haven't tried to understand this breaks the proof
of `multiset.prod_map_add`. Having seen this, I think it's best
to revert this since there's already plenty of surface area in
these changes to break things.

ocfnash committed:
Fix long line

eric-wieser commented on src/algebra/group_with_zero/defs.lean:
```diff
@@ -54,7 +54,11 @@ class no_zero_divisors (M₀ : Type*) [has_mul M₀] [has_zero M₀] : Prop :=
+
+/-- A type `M₀` is a “monoid with zero” if it is a monoid with zero element, and `0` is left
 and right absorbing. -/
 @[protect_proj] class monoid_with_zero (M₀ : Type*) extends monoid M₀, mul_zero_class M₀.
```
Probably a good idea too:
```suggestion
@[protect_proj] class monoid_with_zero (M₀ : Type*) extends monoid M₀, semigroup_with_zero M₀.
```

ocfnash replied:
I'm having a sufficiently painful time with the change to the `monoid` / `add_monoid` ancestors that I'm extremely reluctant to try this too. It just takes so long to test etc.

ocfnash replied:
I'm declaring this out of scope after the trouble I had trying to change the ancestors for `monoid` and `add_monoid` (mostly stemming from a very slow debug loop).

ocfnash committed:
Fix build failure 3 / ?

If this doesn't work, I think I'll give up on inserting `mul_one_class` /
`add_zero_class` into the `monoid` / `add_monoid` definitions.

ocfnash committed:
Fix build failure 4 / ?

ocfnash committed:
Revert changes to `monoid` / `add_monoid` ancestors

The most recent build failure (topology/locally_constant/algebra.lean:52:73)
has persuaded me to abandon this part of the work. It could be done but it's
too much of a diversion / time sink in what is already a tangent.

ocfnash committed:
Improve naming

ocfnash committed:
Move + rename file

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,163 @@
+non_unital_non_assoc_algebra.mul_smul_comm' t a b
+
+/-- The algebra multiplication as a bilinear map. -/
+def mul_as_bilinear {R : Type u₁} {A : Type u₂}
```
We call this `algebra.lmul` in the normal case, right? Can we use the same convention here?

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,163 @@
+instance : inhabited (non_unital_non_assoc_algebra_hom R A B) := ⟨0⟩
+
+lemma coe_injective :
+  function.injective (λ f : non_unital_non_assoc_algebra_hom R A B, show A → B, from f) :=
```
This statement leads to some ugly non-delta reduced expressions when applying it, as I recall. Using `@function.injective _ _ coe_fn` with appropriate underscore values resolves that

ocfnash replied:
Thanks, noted!

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,163 @@
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (A →ₗ[R] B) := ⟨to_linear_map⟩
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (mul_hom A B) := ⟨to_mul_hom⟩
```
Can you add a coercion from `alg_hom`?

eric-wieser replied:
~~Perhaps I'm blind, but I don't see where you've done this.~~

ocfnash committed:
Suggestions from CR

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,170 @@
+⟨λ f, { to_fun    := f,
+        map_add'  := f.map_add,
+        map_smul' := f.map_smul,
+        map_mul'  := f.map_mul, }⟩
```
I suspect (but haven't tested) that this works:
```suggestion
⟨λ f, { to_fun    := f, ..f }⟩
```

ocfnash replied:
Unfortunately there is no `alg_hom.map_smul'` but we can get half way there so I'll commit what we can.

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,170 @@
+
+variables {R A B C}
+
+instance {R A B} [comm_semiring R] [ring A] [ring B] [algebra R A] [algebra R B] :
```
Does this infer the correct universes for `R A B`?

ocfnash replied:
Oh, I'm not at all sure. I didn't even know it was possible to write what I wrote! I can hardly believe this was me.

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,170 @@
+        map_smul' := f.map_smul,
+        map_mul'  := f.map_mul, }⟩
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (A →ₗ[R] B) := ⟨to_linear_map⟩
+
+instance : has_coe (non_unital_non_assoc_algebra_hom R A B) (mul_hom A B) := ⟨to_mul_hom⟩
```
can you add a pair of lemmas of the form `to_mul_hom_eq_coe : f.to_mul_hom = ↑f` here?

ocfnash replied:
Sure

ocfnash committed:
Suggestions' from CR

kim-em commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,176 @@
+
+@[priority 200] -- see Note [lower instance priority]
+instance semiring.to_non_unital_non_assoc_semiring [semiring A] : non_unital_non_assoc_semiring A :=
+{ ..(infer_instance : semiring A) }
```
This belongs in the other file.

kim-em replied:
You can also write `‹semiring α›` rather than `(infer_instance : semiring A)` if you prefer.

ocfnash replied:
Thanks, done. Some resulting build failures which I will resolve shortly.

eric-wieser commented on src/data/equiv/mul_add.lean:
```diff
@@ -34,6 +34,16 @@ equiv, mul_equiv, add_equiv
+  map_mul' := λ x y,
+    calc g (x * y) = g ((f (g x)) * (f (g y))) : by { conv_lhs { rw [←h₂ x, ←h₂ y], }, }
+               ... = g (f ((g x) * (g y))) : by rw f.map_mul
+               ... = (g x) * (g y) : (h₁ _), }
```
can shorten both the length of tactics and the number of parens here:
```suggestion
    calc g (x * y) = g (f (g x) * f (g y)) : by rw [h₂ x, h₂ y]
               ... = g (f (g x * g y)) : by rw f.map_mul
               ... = g x * g y : h₁ _, }
```

ocfnash replied:
Thanks, done.

ocfnash committed:
Move instance and create some others.

Perhaps better to change `semiring` ancestors but it's a slow debug cycle
so I'll start by seeing if this builds.

ocfnash committed:
Add comments about pending refactoring work

ocfnash committed:
Update src/data/equiv/mul_add.lean

Co-authored-by: Eric Wieser wieser.eric@gmail.com

eric-wieser commented on src/algebra/ring/basic.lean:
```diff
@@ -116,6 +116,18 @@ class semiring (α : Type u) extends add_comm_monoid α, monoid_with_zero α, di
+
+@[priority 200] -- see Note [lower instance priority]
+instance semiring.non_assoc_semiring : non_assoc_semiring α :=
+{ ..‹semiring α› }
```
Isn't the default priority for `extends` 100 (eg, see `#print semiring.to_add_comm_monoid`). If there's a reason for using a different 200 here, can you include it in a comment?

ocfnash replied:
Oh sorry I was being stupid. These priority settings are here only because I was mindlessly copying from the definition of `algebra.semimodule`. I've just dropped them. If work is quiet this morning I'll try and resolve the build failures too.

eric-wieser replied:
To be clear; there _should_ be priority settings here, because you're trying to emulate `extends`, and `extends` puts priority 100 on conversion between the derived and base class. My comment was that choosing `200` was odd, not that the presence of `priority` at all was odd.

ocfnash replied:
I've literally just realised this :-)

ocfnash replied:
Doing this all too quickly. Fix incoming!

ocfnash committed:
Drop spurious priority settings

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,182 @@
+
+@[simp] lemma map_smul (f : non_unital_non_assoc_algebra_hom R A B) (c : R) (x : A) :
+  f (c • x) = c • f x :=
+linear_map.map_smul (f : A →ₗ[R] B) c x
```
Untested, but shorter as:
```suggestion
f.to_linear_map.map_smul c x
```

ocfnash replied:
Is this definitely preferable? Maybe I'm being ridiculous used a coercion because it doesn't require me to be opinionated about _how_ to turn `f` into a linear map whereas if I invoke `.to_linear_map` I require that such a method exists. Happy to change it.

eric-wieser replied:
Well, it's a proof so the actual definition you use doesn't really matter. For this specific case, `to_linear_map` is easier to type, shorter, and avoids a typeclass search.

ocfnash replied:
I get that it's a proof so who cares about the term but I'm imagining a hypothetical world in which `non_unital_non_assoc_algebra_hom` still has a coercion to a linear map but by a method of a different name.

In any case, I'll change it: definitely better to avoid typeclass search.

ocfnash committed:
Add missing priorities

Incorrectly dropped instead of adjusted in previous commit

eric-wieser commented on src/algebra/group/defs.lean:
```diff
@@ -174,15 +174,32 @@ theorem mul_left_inj (a : G) {b c : G} : b * a = c * a ↔ b = c :=
 (zero_add : ∀ a : M, 0 + a = a) (add_zero : ∀ a : M, a + 0 = a)
+
 attribute [to_additive] monoid
+attribute [to_additive] mul_one_class
```
This should probably appear before the definition of `monoid`.

eric-wieser commented on src/algebra/group_with_zero/defs.lean:
```diff
@@ -54,7 +54,11 @@ class no_zero_divisors (M₀ : Type*) [has_mul M₀] [has_zero M₀] : Prop :=
 
-/-- A type `M` is a “monoid with zero” if it is a monoid with zero element, and `0` is left
+/-- A type `S₀` is a "semigroup with zero” if it is a semigroup with zero element, and `0` is left
+and right absorbing. -/
```
```suggestion
and right absorbing.

TODO: add the missing conversion from `monoid_with_zero` to `semigroup_with_zero`. -/
```
Unless you want to just add now, in which case do that.

ocfnash replied:
I'll take the TODO :-)

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,182 @@
+  has_coe (A →ₐ[R] B) (non_unital_non_assoc_algebra_hom R A B) :=
+⟨λ f, { to_fun    := f,
+        map_smul' := f.map_smul,
+        ..f, }⟩
```
I think it maybe makes sense to move this to `algebra/algebra/basic`, and swap the direction of the import?

ocfnash replied:
I'm happy to do this if you like but my thought was that by keeping this stuff here, it should be slightly easier to drop what is no longer necessary once `non_unital_non_assoc_algebra` becomes redundant after the refactor of `algebra`.

As I say, happy to change it if you still think worth doing.

eric-wieser replied:
I guess it probably makes things more painful for you since the import graph becomes drastically longer. So ok to leave here, but maybe it makes sense to group all the statements about `algebra` at the bottom of the file; since morally algebra is structure "on top of" `non_unital_non_assoc_algebra`.

ocfnash committed:
Update src/algebra/group_with_zero/defs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Move attribute

ocfnash committed:
Explicitly invoke `to_linear_map` and `to_mul_hom` in simp lemma proofs of `non_unital_non_assoc_algebra_hom`

This has the advantage over the previous of avoiding a typeclass search.

ocfnash committed:
Group statements about `algebra` together at end of file.

kim-em commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,190 @@
+This files defines non-unital, non-associative algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Uniticity, associativity and the legacy `algebra` class.
```
```suggestion
## Unitality, associativity and the legacy `algebra` class.
```

ocfnash replied:
Thanks 🤦

ocfnash committed:
Try to fix build

The breakage is due to the instances placed in `algebra/ring/basic` in
commit 4601470cc200637632d0a498a7fce08ad84d06a3.

I'm yet sure of any of the following:
 1. If the fixes here are sufficient
 2. Why the instances are apparently causing the elaborator to fail
 3. Whether the changes here are the right way to address this issue

I'll wait and see the answer to 1 and then try to understand better.

ocfnash committed:
Fix bad spelling

ocfnash committed:
Add missing instance priority

eric-wieser commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,191 @@
+@[protect_proj]
+class non_unital_non_assoc_algebra [semiring R] [non_unital_non_assoc_semiring A]
+  extends semimodule R A :=
+(smul_mul_assoc' : ∀ (t : R) (a b : A), (t • a) * b = t • (a * b))
+(mul_smul_comm' : ∀ (t : R) (a b : A), a * (t • b) = t • (a * b))
```
I think this made it into a zulip message, but is worth recording in this PR too: these two conditions are respectively `[is_scalar_tower R A A]` and (modulo symmetry) `[smul_comm_class R A A]`.

eric-wieser replied:
At least, it would be if we added `has_mul.to_has_scalar`, since `monoid.to_mul_action` can't apply here

ocfnash replied:
Yep, hadn't forgotten but good to get this reminder.

ocfnash replied:
I just might take advantage of this approach so I'll record the following snippet for emphasis:
```lean
import linear_algebra.basic

variables (R A : Type*) [semiring R] [non_unital_non_assoc_semiring A] [module R A]
variables (t : R) (a b : A)

example [smul_comm_class R A A] : a * (t • b) = t • (a * b) := (smul_comm t a b).symm

example [is_scalar_tower R A A] : (t • a) * b = t • (a * b) := smul_assoc t a b
```

Incidentally, the scalar action of `A` on itself is obtained from the combination of:

- `non_unital_non_assoc_semiring.to_mul_zero_class`
- `mul_zero_class.to_smul_with_zero`
- `smul_with_zero.to_has_scalar`

eric-wieser replied:
I forgot that `smul_with_zero.to_has_scalar` applied here!

kbuzzard replied:
Wait -- so doesn't this just mean that either one doesn't need the class at all, or that one can just make the class by extending other classes?

We had this sort of thing with number fields too -- 

```
import linear_algebra.finite_dimensional

variables (K : Type*) [field K] [char_zero K] [finite_dimensional ℚ K]
```
is mathematically "let K be a number field" so now you start wondering whether you even need a number field typeclass (and indeed I don't think we have one)

ocfnash replied:
Exactly right! Hence my remark at the top of #7847 about it being a plausible replacement for this class.

ocfnash replied:
(although we do face the problem of "exponential term size" if we go too far this way)

ocfnash committed:
Merge branch 'master' into non_assoc_algebras

ocfnash committed:
Clean up after merge

ocfnash committed:
Fix: `semimodule` → `module`

kbuzzard commented on src/algebra/algebra/non_unital.lean:
```diff
@@ -0,0 +1,191 @@
+
+lemma smul_mul_assoc (t : R) (a b : A) :
+  (t • a) * b = t • (a * b) :=
+non_unital_non_assoc_algebra.smul_mul_assoc' t a b
```
Often the point of having a primed field in a structure and an unprimed one directly afterwards is that the unprimed one can take advantage of things like coercions (e.g. it can refer to `⇑f` rather than `f.to_fun`) but here it seems to me that the primed and unprimed versions are exactly the same even up to the binders. Is it clear that we need the primed ones?

ocfnash replied:
Agreed. Fortunately this PR is (hopefully) going to be closed in favour of #7847 so this detail can be forgotten. It's old and I'd guess I wrote it before I understood this convention.

kbuzzard reviewed (commented):
This looks fine to me, but I am not an expert in the subtleties of these things. I've left a couple of other comments.
